From e925ebc994c72b830345c2cb9a8a7c6595292eac Mon Sep 17 00:00:00 2001
From: Alexander Martinz <amartinz@shiftphones.com>
Date: Thu, 28 Mar 2019 08:37:54 +0100
Subject: [PATCH 2/2] Camera2: also check for and request
 WRITE_EXTERNAL_STORAGE permission

We are asking for READ permissions already, but not for WRITE.
In case we do not get all specified permissions pregranted and the app
asks for them again (eg if installing a development version or user
plays with permissions), the app is not functional and crashes when
taking pictures or trying to delete existing ones.

To fix this, also check for WRITE permissions and if not granted, request
them.

Change-Id: I50b3c44375b5dfa2aa92031a17dc5793dede15eb
Signed-off-by: Alexander Martinz <amartinz@shiftphones.com>
---
 src/com/android/camera/CameraActivity.java    |  3 +-
 .../android/camera/PermissionsActivity.java   | 33 +++++++++++--------
 2 files changed, 21 insertions(+), 15 deletions(-)

diff --git a/src/com/android/camera/CameraActivity.java b/src/com/android/camera/CameraActivity.java
index 3ff843b4f..d292f26f5 100644
--- a/src/com/android/camera/CameraActivity.java
+++ b/src/com/android/camera/CameraActivity.java
@@ -1918,7 +1918,8 @@ public class CameraActivity extends QuickActivity
 
         if (checkSelfPermission(Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED &&
                 checkSelfPermission(Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED &&
-                checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
+                checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED &&
+                checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
             mHasCriticalPermissions = true;
         } else {
             mHasCriticalPermissions = false;
diff --git a/src/com/android/camera/PermissionsActivity.java b/src/com/android/camera/PermissionsActivity.java
index 5617fe599..d6e89a9ad 100644
--- a/src/com/android/camera/PermissionsActivity.java
+++ b/src/com/android/camera/PermissionsActivity.java
@@ -1,10 +1,8 @@
 package com.android.camera;
 
 import android.Manifest;
-import android.app.Activity;
-import android.app.Dialog;
 import android.app.AlertDialog;
-import android.app.KeyguardManager;
+import android.app.Dialog;
 import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.DialogInterface;
@@ -15,6 +13,7 @@ import android.os.Bundle;
 import android.view.KeyEvent;
 import android.view.Window;
 import android.view.WindowManager;
+
 import com.android.camera.app.CameraServicesImpl;
 import com.android.camera.debug.Log;
 import com.android.camera.settings.Keys;
@@ -96,33 +95,34 @@ public class PermissionsActivity extends QuickActivity {
         unregisterReceiver(mShutdownReceiver);
     }
 
+    private boolean isPermissionGranted(final String permission) {
+        return checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED;
+    }
+
     private void checkPermissions() {
-        if (checkSelfPermission(Manifest.permission.CAMERA)
-                != PackageManager.PERMISSION_GRANTED) {
+        if (!isPermissionGranted(Manifest.permission.CAMERA)) {
             mNumPermissionsToRequest++;
             mShouldRequestCameraPermission = true;
         } else {
             mFlagHasCameraPermission = true;
         }
 
-        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO)
-                != PackageManager.PERMISSION_GRANTED) {
+        if (!isPermissionGranted(Manifest.permission.RECORD_AUDIO)) {
             mNumPermissionsToRequest++;
             mShouldRequestMicrophonePermission = true;
         } else {
             mFlagHasMicrophonePermission = true;
         }
 
-        if (checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE)
-                != PackageManager.PERMISSION_GRANTED) {
-            mNumPermissionsToRequest++;
+        if (!isPermissionGranted(Manifest.permission.READ_EXTERNAL_STORAGE) ||
+                !isPermissionGranted(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+            mNumPermissionsToRequest += 2;
             mShouldRequestStoragePermission = true;
         } else {
             mFlagHasStoragePermission = true;
         }
 
-        if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION)
-                != PackageManager.PERMISSION_GRANTED) {
+        if (!isPermissionGranted(Manifest.permission.ACCESS_FINE_LOCATION)) {
             mNumPermissionsToRequest++;
             mShouldRequestLocationPermission = true;
         }
@@ -159,6 +159,10 @@ public class PermissionsActivity extends QuickActivity {
             permissionsToRequest[permissionsRequestIndex] = Manifest.permission.READ_EXTERNAL_STORAGE;
             mIndexPermissionRequestStorage = permissionsRequestIndex;
             permissionsRequestIndex++;
+
+            permissionsToRequest[permissionsRequestIndex] = Manifest.permission.WRITE_EXTERNAL_STORAGE;
+            mIndexPermissionRequestStorage = permissionsRequestIndex;
+            permissionsRequestIndex++;
         }
         if (mShouldRequestLocationPermission) {
             permissionsToRequest[permissionsRequestIndex] = Manifest.permission.ACCESS_FINE_LOCATION;
@@ -195,8 +199,9 @@ public class PermissionsActivity extends QuickActivity {
             }
         }
         if (mShouldRequestStoragePermission) {
-            if (grantResults.length > 0 && grantResults[mIndexPermissionRequestStorage] ==
-                    PackageManager.PERMISSION_GRANTED) {
+            if (grantResults.length >= 2 &&
+                    grantResults[mIndexPermissionRequestStorage - 1] == PackageManager.PERMISSION_GRANTED &&
+                    grantResults[mIndexPermissionRequestStorage] == PackageManager.PERMISSION_GRANTED) {
                 mFlagHasStoragePermission = true;
             } else {
                 handlePermissionsFailure();
-- 
2.17.1

